### ISSUE TITLE
Invalid 'Same-type constraint is recursive' error

### ISSUE THREAD

**Issue by ffried** [apple/swift#62314]
**Description**
The following code results in an unexpected error:

```swift
protocol X {}

protocol P {
    associatedtype V
}

protocol A {
    associatedtype I
}

protocol B {
    associatedtype M: A
    associatedtype IP: P where IP.V == M.I
}

struct S<V>: P where V: X {}

extension B where M.I: X, IP == S<M.I> {}
```

Error:
```
recursive-constraint.swift:18:13: error: same-type constraint 'Self.IP' == 'S<Self.IP.V>' is recursive
extension B where M.I: X, IP == S<M.I> {}
```

I don't see how the constraints in the code are recursive. However, I think the issue stems from what the compiler interprets here, since the constraint in the error message indeed _is_ recursive: `'Self.IP' == 'S<Self.IP.V>'`. I don't know, though, how it gets from 
`IP == S<M.I>` (aka. `Self.IP == S<Self.M.I>`) to `'Self.IP' == 'S<Self.IP.V>`.

The only hint into that direction would be the `where IP.V == M.I` on the associated type `B.IP`, but even leaving away that constraint leads to the same error.

**Steps to reproduce**
Place the code above into a file (e.g. `recursive-constraint.swift`) and run it with `swift recursive-constraint.swift`.

**Expected behavior**
The compilation completes successfully.

**Environment**
- Swift compiler version info: swift-driver version: 1.62.15 Apple Swift version 5.7.1 (swiftlang-5.7.1.135.3 clang-1400.0.29.51) Target: arm64-apple-macosx13.0
- Xcode version info: Xcode 14.1 Build version 14B47b
- Deployment target: macOS 13.0

**Comment by AnthonyLatsis** [comment]
> I don't know, though, how it gets from `IP == S<M.I>` (aka. `Self.IP == S<Self.M.I>`) to `Self.IP == S<Self.IP.V>`.

`IP == S<M.I>` (from the `where` clause on the extension) + `IP.V == M.I` (from the `where` clause on the associated type `IP`) → `IP == S<IP.V>`, which is indeed recursive. There is no substitution that would satisfy this requirement.

**Comment by ffried** [comment]
Maybe I'm just unable to wrap my head around the actual problem, but in the following example implementation, the constraints would all be satisfied, wouldn't they?

```swift
struct T: X {}

struct U: A {
    typealias I = T
}

struct V: B {
    typealias M = U
    typealias IP = S<T>
}
```

Either way, as I noted above, the `where` clause on the associated type can be left away and it still results in the same error.

**Comment by AnthonyLatsis** [comment]
> Either way, as I noted above, the `where` clause on the associated type can be left away and it still results in the same error.

The requirement `IP == S<M.I>` implies that `IP.V == S<M.I>.V`, and `S<M.I>.V == M.I` per the conformance of `S` to `P`, so either way we end up with `IP == S<IP.V>`.

> in the following example implementation, the constraints would all be satisfied, wouldn't they?

Heh, interesting! I stand corrected. @slavapestov Are we equipped to unseal this form of constraint?

**Comment by slavapestov** [comment]
> Either way, as I noted above, the where clause on the associated type can be left away and it still results in the same error.

This is a bit of a weird case.

```swift
extension B where Self.M.I: X, IP == S<M.I> {}
```

If `IP == S<M.I>`, then `IP.V == M.I` because the generic parameter V witnesses the associated type V of P, and V is bound to M.I in the same-type requirement. but according to our canonical type order, IP.V precedes M.I, so the canonical type of S<M.I> becomes S<IP.V>, thus the minimized signature is `IP == S<IP.V>, IP.V == M.I`.

**Comment by AnthonyLatsis** [comment]
> The issue is that the conformance requirement `IP: P` comes from a protocol so it cannot be minimized away, thus `IP.V` always exists and precedes `M.I`. I'm not sure if there's any way we can change this.

So, I take it you mean we may not be able to address the diagnostics QoI part, but can we allow this kind of constraint?

**Comment by slavapestov** [comment]
Yeah, in theory we can allow something like `S == G<S.T>` as long as there's a `S: P` conformance implied somewhere else (and not directly stated in this signature). It would require other representational changes though.
